### PR TITLE
LDpred2 fixes

### PR DESCRIPTION
**Pull request by deepchocolate**
Fixes # Not Sure

Changes proposed in this pull request:

Changes to [LDpred2.R](https://github.com/comorment/containers/blob/main/usecases/LDpred2/ldpred2.R):
- Summary statistics
	- Added exception when necessary columns are not found in sumstatsfile
	- Made names of columns in sumstatfiles case insensitive
	- Added exception when no information on effective sample size was available
	- Harmonisation of alleles (uppercase conversion if lowercase in sumstats). LDpred ignores such differences when matching on chromosome, alleles and position or rsid (you are just left with 0 SNPs matched).
- Genotype data
	- Separated the creation of the backingfile (bigSNPr .rds & .bk) from the main script. This is now handled by createBackingFile.R. This makes it more transparent to the end-user which files are created and provides a cleaner role of the main script.
	- A warning will be issued if provided genotype data contain missing values. To prevent failure **mean** imputation is applied (similar to prsice). An option is added so that the user can specifiy a different method (one given by `bigsnpr::fastImputeSimple`). More advanced methods I think should be the task for a separate script.

- Phenotypes
	- Added option of using phenotype available in plinks fam file
	- Fixed bug of hardcoded phenotype (Height)
	- Removed diagnostics of PGS when no phenotype is provided.

- Fixed conversion from physical to genomic position using maps from 1000 genomes.
	- The user can specify an option that should be a directory in which the raw files from the [1000 genomes maps repository](https://github.com/joepickrell/1000-genomes-genetic-maps) should be stored. Otherwise ldpred tries to download these files (generating a failure if no internet is available).
	- Another option can be used to specify which map will be used, hapmap or OMNI (don't know the difference).

- Output
	- Added option of naming the column with the outputted PGS in the outputted file

- Other
	- Added more parameters to LDpred models (window size, hyper parameter p)
	- Added option of setting a seed for reproducibility
	- Added exception if invalid option is passed to `--ldpred-mode`
	- Added script for test runs in `tests/ldpred2/`

### Notes
 - Chromosome 23 is omitted as the genomic maps are not available. Beyond my knowledge to deal with so the script just leaves a notification that it was omitted.
 - LD estimation: If I remember correctly, we decided to take this out of the script or add an option of providing LD from a file (as is option provided by the LDpred2 tutorial). I looked at that data and it was not straightforward to use the file they provided, so I think that issue is for the next round.

## Before submitting

- [x] I've read and followed all steps in the [Making a pull request](https://github.com/comorment/containers/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/comorment/containers/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [x] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

### CODE REVIEW

**Comment by espenhgn** [comment]
Hi @deepchocolate, just a comment/question while this is in development :-)
What exact version of `bigsnpr` are you assuming/basing the LDpred2.R analysis script on? 
At the moment the version we'll include (as soon as a new `r.sif` build is pushed) is [v1.11.4](https://github.com/comorment/containers/blob/c2f3348342e80c12a7c337f93c767a85ecef856d/src/scripts/R/cran.R#L4) from CRAN. 
This LDpred2 [tutorial](https://privefl.github.io/bigsnpr/articles/LDpred2.html#installation) implies that there are version differences for the `auto` method, so we just need to make sure that we don't inadvertently introduce some discrepancies.

**Comment by deepchocolate** [comment]
> Hi @deepchocolate, just a comment/question while this is in development :-) What exact version of `bigsnpr` are you assuming/basing the LDpred2.R analysis script on? At the moment the version we'll include (as soon as a new `r.sif` build is pushed) is [v1.11.4](https://github.com/comorment/containers/blob/c2f3348342e80c12a7c337f93c767a85ecef856d/src/scripts/R/cran.R#L4) from CRAN. This LDpred2 [tutorial](https://privefl.github.io/bigsnpr/articles/LDpred2.html#installation) implies that there are version differences for the `auto` method, so we just need to make sure that we don't inadvertently introduce some discrepancies.

Good question! I was about to ask you if you had updated the `r.sif` file, because I got an error that bigsnpr wasn't available there when trying to use that container. So, I have continued using the container I created when developing. The container you created uses `1.11.4`? I have been using `1.11.6`. Seems to be the most recent on cran (Nov 7 22)

**Comment by espenhgn** [comment]
> 



> > Hi @deepchocolate, just a comment/question while this is in development :-) What exact version of `bigsnpr` are you assuming/basing the LDpred2.R analysis script on? At the moment the version we'll include (as soon as a new `r.sif` build is pushed) is [v1.11.4](https://github.com/comorment/containers/blob/c2f3348342e80c12a7c337f93c767a85ecef856d/src/scripts/R/cran.R#L4) from CRAN. This LDpred2 [tutorial](https://privefl.github.io/bigsnpr/articles/LDpred2.html#installation) implies that there are version differences for the `auto` method, so we just need to make sure that we don't inadvertently introduce some discrepancies.
> 
> Good question! I was about to ask you if you had updated the `r.sif` file, because I got an error that bigsnpr wasn't available there when trying to use that container. So, I have continued using the container I created when developing. The container you created uses `1.11.4`? I have been using `1.11.6`. Seems to be the most recent on cran (Nov 7 22)

Hi! No, we have not pushed any updated container builds to the main branch as the corresponding Dockerfiles have seen some frequent changes recently. We should rather push updated container builds for major updates (new functionality and such), in a release v0.2 or similar due to their size. I can push an updated build onto another temporary branch though if you give me a moment.

**Comment by deepchocolate** [comment]
> > 
> 
> > > Hi @deepchocolate, just a comment/question while this is in development :-) What exact version of `bigsnpr` are you assuming/basing the LDpred2.R analysis script on? At the moment the version we'll include (as soon as a new `r.sif` build is pushed) is [v1.11.4](https://github.com/comorment/containers/blob/c2f3348342e80c12a7c337f93c767a85ecef856d/src/scripts/R/cran.R#L4) from CRAN. This LDpred2 [tutorial](https://privefl.github.io/bigsnpr/articles/LDpred2.html#installation) implies that there are version differences for the `auto` method, so we just need to make sure that we don't inadvertently introduce some discrepancies.
> > 
> > 
> > Good question! I was about to ask you if you had updated the `r.sif` file, because I got an error that bigsnpr wasn't available there when trying to use that container. So, I have continued using the container I created when developing. The container you created uses `1.11.4`? I have been using `1.11.6`. Seems to be the most recent on cran (Nov 7 22)
> 
> Hi! No, we have not pushed any updated container builds to the main branch as the corresponding Dockerfiles have seen some frequent changes recently. We should rather push updated container builds for major updates (new functionality and such), in a release v0.2 or similar due to their size. I can push an updated build onto another temporary branch though if you give me a moment.

Ah ok! But no rush, I can wait. I doubt there are any dramatic changes between these versions

**Comment by espenhgn** [comment]
@deepchocolate; An updated [r.sif](https://github.com/comorment/containers/blob/tmp_new_r_sif_bld/singularity/r.sif) build is pushed via the branch [tmp_new_r_sif_bld](https://github.com/comorment/containers/tree/tmp_new_r_sif_bld). Hope it works on your end!

**Comment by deepchocolate** [comment]
@ofrei @espenhgn LDpred uses genetic maps from [1000 Genomes](https://github.com/joepickrell/1000-genomes-genetic-maps) for estimating LD. These are not available in the repo, right?

It's used by a function `snp_asGeneticPos(CHR, POS, directory, rsid, type)` to convert from physical position in basepairs to genomic position in centiMorgans. `type` can be either "OMNI" or "hapmap", and `directory` a folder were the unzipped files are located (otherwise LDpred tries to download them). The option `rsid` allows you tu use these instead, but how that works is not illustrated by the tutorial and I can't say if there are any drawbacks to using that instead or whether you will still need the maps. As far as I understand the files from 1000 genomes need to be named as in the repo as there is no option of specifying a different naming.

**Comment by espenhgn** [comment]
> 



> @ofrei @espenhgn LDpred uses genetic maps from [1000 Genomes](https://github.com/joepickrell/1000-genomes-genetic-maps) for estimating LD. These are not available in the repo, right?
> 
> It's used by a function `snp_asGeneticPos(CHR, POS, directory, rsid, type)` to convert from physical position in basepairs to genomic position in centiMorgans. `type` can be either "OMNI" or "hapmap", and `directory` a folder were the unzipped files are located (otherwise LDpred tries to download them). The option `rsid` allows you tu use these instead, but how that works is not illustrated by the tutorial and I can't say if there are any drawbacks to using that instead or whether you will still need the maps. As far as I understand the files from 1000 genomes need to be named as in the repo as there is no option of specifying a different naming.

Hi. I don't think these exact files are included. There are some converted files from 1000 Genomes in https://github.com/comorment/containers/tree/main/reference/magma/g1000_eur but it does look like their origin is different. It's a bit annoying, but if we have to include this 912MB dataset, we should. It won't be an option to download it automatically on e.g., TSD anyway. 

This [tutorial](https://privefl.github.io/bigsnpr/articles/LDpred2.html#which-set-of-variants-to-use) suggests only a GWAS sumstats file and an updated LD (correlation) matrix file from HapMap3 is required for LDpred2, so I'm a bit confused why the package would download something else in addition, to be frank.

**Comment by ofrei** [comment]
@espen @deepchocolate  I agree with including genomic maps, despite them taking 912 MB . But we should also consider including the file(s) needed as --file-gene-corr argument of the ldpred2.R script.

**Comment by deepchocolate** [comment]
> @espen @deepchocolate I agree with including genomic maps, despite them taking 912 MB . But we should also consider including the file(s) needed as --file-gene-corr argument of the ldpred2.R script.

I have added an argument so that the user can specify where these files are located

**Comment by espenhgn** [comment]
> 

Sounds good! Btw., I sit here now with @ofrei and wonder if this can be merged before tomorrow's meeting, or if you need more time before this is good to go?

**Comment by deepchocolate** [comment]
> > 
> 
> Sounds good! Btw., I sit here now with @ofrei and wonder if this can be merged before tomorrow's meeting, or if you need more time before this is good to go?

I hope I'm done with the last edits today!

But I noted that it's probably a bad design to do both the LD and the scoring in one script as the LD takes the bulk of the computational time. I'm guessing you can reuse the same LD estimates for several different scores assuming that the majority of the SNPs overlap. But maybe that can be saved for later

**Comment by deepchocolate** [comment]
@espenhgn @ofrei I'll push the last changes today.

I looked up more in detail how PRsice2 handles missing genotypes. PRsice2 argument `--missing` can take on `MEAN_IMPUTE`, `SET_ZERO`, and `CENTER` (mean zero). It appears that verion 1.2.1 of bigSNPr has added the option to impute by zero in `snp_fastImputSimple`: https://www.rdocumentation.org/packages/bigsnpr/versions/1.2.1/topics/snp_fastImputeSimple

**Comment by deepchocolate** [comment]
@espenhgn I switched to using the repo containers now. I've encountered two problems

1) Running the container generates a warning about locale (also everytime a parallell computation is initiated).
```
During startup - Warning messages:
1: Setting LC_TIME failed, using "C" 
2: Setting LC_MONETARY failed, using "C" 
3: Setting LC_PAPER failed, using "C" 
4: Setting LC_MEASUREMENT failed, using "C"
```
Is there someting in my local environment that triggers this? I had the same issue with the container I created myself, but I fixed that by adding the below in the singularity definition file:
```
%environment
	export LC_ALL=C
```

2) There is some problem with the bigsnpr and parallellism:
```
Error: Two levels of parallelism are used. See `?assert_cores`.
Execution halted
```
`assert_cores` says that
```
It also checks if two levels of parallelism are used, i.e. having ncores larger than 1, and having a parallel BLAS enabled by default. You could remove this check by setting options(bigstatsr.check.parallel.blas = FALSE).

We instead recommend that you disable parallel BLAS by default by adding try(bigparallelr::set_blas_ncores(1), silent = TRUE) to your .Rprofile (with an empty line at the end of this file) so that this is set whenever you start a new R session. You can use usethis::edit_r_profile() to open your .Rprofile. For this to be effective, you should restart the R session or run options(default.nproc.blas = NULL) once in the current session.

Then, in a specific R session, you can set a different number of cores to use for matrix computations using bigparallelr::set_blas_ncores(), if you know there is no other level of parallelism involved in your code.
```
It seems to work if I add `options(bigstatsr.check.parallel.blas = FALSE)` to the beginning of the script, but I'm not sure if that's the right way to solve it.

**Comment by espenhgn** [comment]
Hi @deepchocolate, 
The `LC_*` issue you see is likely explained in this thread: https://stackoverflow.com/questions/9689104/installing-r-on-mac-warning-messages-setting-lc-ctype-failed-using-c. 
Didn't test this yet, but adding `LC_ALL=C.UTF-8` to the host list of environment variables may fix the warnings. 
Prepending `SINGULARITYENV_` may allow the running container to pick it up. 

This may also work:
```
export LANG=en_US.UTF-8
export LC_ALL=en_US.UTF-8
```

I also encountered the parallelism issue you encountered. I fixed the crash similarly in https://github.com/comorment/containers/blob/main/usecases/LDpred2_tutorial/LDpred2.Rmd by adding the lines
```
## fix "Error: Two levels of parallelism are used. See `?assert_cores`."
options(bigstatsr.check.parallel.blas = FALSE)
options(default.nproc.blas = NULL)
```
after importing `bigsnpr` and `bigreadr`. So if I understand you correctly, this error occurs only with the container we built, and not yours? 
I see we install the ubuntu `libgslcblas0` apt package on top of the rocker verse Docker container (defined at https://github.com/rocker-org/rocker-versioned2) as some R package we install required it, but I don't know yet if this is the exact reason.
We should definitely check that multithreading still works for R scripts. 
To me it looks like `bigsnpr` prefers to handle threading behaviour itself, that is, `ncores=#cores` is a keyword argument to multiple functions, so it seemingly makes sense to use blas in serial mode.

**Comment by deepchocolate** [comment]
> Hi @deepchocolate, The `LC_*` issue you see is likely explained in this thread: https://stackoverflow.com/questions/9689104/installing-r-on-mac-warning-messages-setting-lc-ctype-failed-using-c. Didn't test this yet, but adding `LC_ALL=C.UTF-8` to the host list of environment variables may fix the warnings. Prepending `SINGULARITYENV_` may allow the running container to pick it up.
> 
> This may also work:
> 
> ```
> export LANG=en_US.UTF-8
> export LC_ALL=en_US.UTF-8
> ```
> 
> I also encountered the parallelism issue you encountered. I fixed the crash similarly in https://github.com/comorment/containers/blob/main/usecases/LDpred2_tutorial/LDpred2.Rmd by adding the lines
> 
> ```
> ## fix "Error: Two levels of parallelism are used. See `?assert_cores`."
> options(bigstatsr.check.parallel.blas = FALSE)
> options(default.nproc.blas = NULL)
> ```
> 
> after importing `bigsnpr` and `bigreadr`. So if I understand you correctly, this error occurs only with the container we built, and not yours? I see we install the ubuntu `libgslcblas0` apt package on top of the rocker verse Docker container (defined at https://github.com/rocker-org/rocker-versioned2) as some R package we install required it, but I don't know yet if this is the exact reason. We should definitely check that multithreading still works for R scripts. To me it looks like `bigsnpr` prefers to handle threading behaviour itself, that is, `ncores=#cores` is a keyword argument to multiple functions, so it seemingly makes sense to use blas in serial mode.

1)
I set the `LC_ALL` environment variable in my testing script, which solved it :)

2)
Correct, the error does not appear in the container I built. But if I understand you correctly, I'll proceed by setting these options in the `ldpred2.R` script for now.

**Comment by espenhgn** [comment]
> Correct, the error does not appear in the container I built. But if I understand you correctly, I'll proceed by setting these options in the `ldpred2.R` script for now.

I believe this is "safe". I found my quickfix in another [LDpred2 tutorial](https://choishingwan.github.io/PRS-Tutorial/ldpred/#0-prepare-workspace). There's also some more discussion on this issue at https://github.com/privefl/bigstatsr/issues/90

**Comment by espenhgn** [comment]
For whatever reason, Rocker specify these env variables in their Dockerfiles, see https://github.com/rocker-org/rocker-versioned2/blob/a3af9efeee8e5e4d617b3ec9c6614802f2da41d2/dockerfiles/Dockerfile_r-ver_4.0.5#L10

The question is whether or not we should override this in our Dockerfile. 
The outcome is in anyway clear, that is, the locale is different between containers: 
```
espenh@comorment:/nrec/space/espenh/containers$ singularity run singularity/gwas.sif locale
LANG=C.UTF-8
LANGUAGE=
LC_CTYPE="C"
LC_NUMERIC="C"
LC_TIME="C"
LC_COLLATE="C"
LC_MONETARY="C"
LC_MESSAGES="C"
LC_PAPER="C"
LC_NAME="C"
LC_ADDRESS="C"
LC_TELEPHONE="C"
LC_MEASUREMENT="C"
LC_IDENTIFICATION="C"
LC_ALL=C
espenh@comorment:/nrec/space/espenh/containers$ singularity run singularity/r.sif locale
LANG=en_US.UTF-8
LANGUAGE=
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
```
I suppose we can live with the warnings in this case (if warnings are all there is)

**Comment by ofrei** [comment]
@deepchocolate everything looks great, thanks for putting it all together!
I was able to run the ``run.sh`` from ``tests/ldpred2`` - it is straighforward, and I can use this script to get started.
So I think we can merge. Tomorrow or on Friday I'm planning to give it a try on real data, and if we need further  changes  we can submit them as separate PR.

Here is output I got from ``run.sh``:
```
Testing RDS/backingfile creation

Testing mode inf
WARNING: While bind mounting '/nrec/space/ofrei/github/comorment/ldsc/reference:/REF': destination is already in the mount point list
Loading backingfile: /nrec/space/ofrei/github/comorment/containers/tests/ldpred2/data/EUR.rds 

### Reading summary statistics /REF/examples/prsice2/Height.gwas.txt.gz 
Loaded 529504 SNPs
Filtering SNPs using /REF/hapmap3/w_hm3.justrs 
Retained 138489 out of 529504 
Converting odds-ratio to log scale
138,489 variants to be matched.
8 ambiguous SNPs have been removed.
138,474 variants have been matched; 0 were flipped and 2,081 were reversed.

### Calculating SNP correlation/LD using 1 cores
Converting from physical position to genetic position
NOTE: Omitting chromosome 23 in LD estimation
Using file temp/fileceff27bea2b3c to store matrixes
Chromosome: 1...2...3...4...5...6...7...8...9...10...11...12...13...14...15...16...17...18...19...20...21...22...[1] 0
 num [1:138474] 1.14 1.42 1.59 2.6 1.73 ...

### Running LD score regression
Results: Intercept = 0.6549324 H2 = 0.4710137 

### Starting polygenic scoring
Running LDPRED2 infinitesimal model
Calculating beta inf
Scoring all individuals...Warning message:
Missing genotypes found (N positions=544575). Imputing genotypes by using "mean0" (see bigsnpr::snp_fastImputeSimple).
 

### Writing fam file with PGS and phenotype
```

**Review by ofrei — APPROVED** [review]

**Comment by ofrei** [comment]
@espenhgn I rebased and made a [minor change](https://github.com/comorment/containers/commit/7e1314b38b3f5876209addfb5e320fd24bd683d3) in the CHANGELOG due to rebase, these are in https://github.com/comorment/containers/tree/fix-96-additions-v2 branch. Please merge that branch or "git cherry-pick" that commit.

**Comment by espenhgn** [comment]
hi @deepchocolate. Thanks for updating this! 

For me, the first example in the [README](https://github.com/comorment/containers/tree/deepchocolate/ldpred2_updates/usecases/LDpred2#running-ldpred2-analysis) throws an error:
```
$ bash run_ldpred2.sh
Loading required package: bigstatsr
[1] "EUR.imputed.bed"
[1] TRUE
[1] TRUE
Loading backingfile: EUR.imputed.bed 
Error in readRDS(rdsfile) : unknown input format
Calls: snp_attach -> readRDS
Execution halted
Loading backingfile: EUR.imputed.bed 
Error in readRDS(rdsfile) : unknown input format
Calls: snp_attach -> readRDS
Execution halted
```
I pulled in the current `r.sif`  and `gwas.sif` files from branch `main` for this. 

As for the `<containers>/tests/` directory, I think this directory should exclusively be for files related to the unit testing setup. Sorry if this was not clearly communicated. The gist of it is that the Python `pytest` framework is used to check the different software and analysis in each container. A statement like `py.test -v tests/` from the root directory should run all checks, `py.test -v tests/test_r.py` should test stuff related to the `r.sif` container file, etc.

I could also run the `tests/ldpred2/run.sh` file with the only issue being the warning @ofrei already noted. However, I suggest that the `tests/ldpred2` files be merged or relocated to `usecases/LDpred2`.

**Review by espenhgn — CHANGES_REQUESTED** [review]
For reasons explained in https://github.com/comorment/containers/pull/94#issuecomment-1333448968 I think the files in /tests/ldpred2/ should be relocated within the `usecases` directory

**Comment by deepchocolate** [comment]
@ofrei Great, thanks!

@espenhgn Ok! I'll add a README and move the test directory. Does it sound good to place it in `usecases/LDpred2/tests/`?
I'll fix the error in the README.md

**Comment by espenhgn** [comment]
> @ofrei Great, thanks!
> 
> @espenhgn Ok! I'll add a README and move the test directory. Does it sound good to place it in `usecases/LDpred2/tests/`? I'll fix the error in the README.md

Thanks. I think you can move the `tests/ldpred2/` files into a new folder `usecases/LDpred2_example`

**Comment by deepchocolate** [comment]
@espenhgn I think I've addressed these things now. I also updated the README to remove the imputation part as that is now handled by the `LDpred2.R` and added creation of the RDS file. 

But there still is the issue with the genomic maps as `--dir-genetic-maps` is a mandatory flag without internet connection. Shall I add something about this in the README and `run_ldpred2.sh`?

File: usecases/LDpred2/ldpred2.R
Comment (espenhgn) [inline usecases/LDpred2/ldpred2.R:243]: This qplot function in qqplot2 3.4.0 is deprecated, see https://ggplot2.tidyverse.org/reference/qplot.html). A warning is thrown, so not an immediate concern. I see the version I specified to be installed in the `r.sif` build is 3.3.5, however, 3.4.0 is installed for some reason.

**Comment by espenhgn** [comment]
> @espenhgn I think I've addressed these things now. I also updated the README to remove the imputation part as that is now handled by the `LDpred2.R` and added creation of the RDS file.
> 
> But there still is the issue with the genomic maps as `--dir-genetic-maps` is a mandatory flag without internet connection. Shall I add something about this in the README and `run_ldpred2.sh`?

Hi @deepchocolate, I think it looks good! If you can mention the issue with `--dir-genetics-maps` in the README that would be good, I agree. I encountered a minor issue with the `run_ldpred2.sh` script. PR #103 fixes the issue on my end anyway.

File: usecases/LDpred2/ldpred2.R
Comment (espenhgn) [reply usecases/LDpred2/ldpred2.R:243]: Opened issue #104 for this.

Comment (deepchocolate) [reply usecases/LDpred2/ldpred2.R:243]: I can rewrite the plotting code using pure ggplot, should be quite simple

Comment (espenhgn) [reply usecases/LDpred2/ldpred2.R:243]: > I can rewrite the plotting code using standard ggplot, should be quite simple

Ok, that'd be more future proof

**Comment by ofrei** [comment]
@deepchocolate I'm eager to get started with testing LDpred2.R using simulated and real data. Could you please merge this pull request, and then follow up on remaining changes you'd like to implement in a separate PR? I really don't need this to be well tested - the point is just to have some a well defined version in the main branch that I'm going to test.

**Comment by deepchocolate** [comment]
> @deepchocolate I'm eager to get started with testing LDpred2.R using simulated and real data. Could you please merge this pull request, and then follow up on remaining changes you'd like to implement in a separate PR? I really don't need this to be well tested - the point is just to have some a well defined version in the main branch that I'm going to test.

I'm about to push the last changes, but I'm blocked from merging until @espenhgn approves

File: usecases/LDpred2/ldpred2.R
Comment (espenhgn) [reply usecases/LDpred2/ldpred2.R]: I'm marking this as resolved if it is blocking merging, as it is not a high priority

**Review by espenhgn — APPROVED** [review]
Thanks a lot @deepchocolate. I'll merge this PR in a moment.

**Comment by espenhgn** [comment]
> > @deepchocolate I'm eager to get started with testing LDpred2.R using simulated and real data. Could you please merge this pull request, and then follow up on remaining changes you'd like to implement in a separate PR? I really don't need this to be well tested - the point is just to have some a well defined version in the main branch that I'm going to test.
> 
> I'm about to push the last changes, but I'm blocked from merging until @espenhgn approves

Approved!

**Pull request merged**
